mongo/integration/cmd_monitoring_helpers_test.go

```diff
@@ -31,57 +37,70 @@ func numberFromValue(mt *mtest.T, val bson.RawValue) int64 {
-		return
+		if actual.Type == bson.TypeNull {
+			return fmt.Errorf("expected non-null value for key %s, got null", key)
+		}
+		return nil
 	}
 
 	aInt := numberFromValue(mt, actual)
-	assert.Equal(mt, eInt, aInt, "value mismatch for key %s; expected %v, got %v", key, expected, actual)
+	if eInt != aInt {
+		return fmt.Errorf("value mismatch for key %s; expected %s, got %s", key, expected, actual)
+	}
+	return nil
 }
 
 // compare BSON values and fail if they are not equal. the key parameter is used for error strings.
 // if the expected value is a numeric type (int32, int64, or double) and the value is 42, the function only asserts that
 // the actual value is non-null.
-func compareValues(mt *mtest.T, key string, expected, actual bson.RawValue) {
+func compareValues(mt *mtest.T, key string, expected, actual bson.RawValue) error {
 	mt.Helper()
 
 	switch expected.Type {
 	case bson.TypeInt32, bson.TypeInt64, bson.TypeDouble:
-		compareNumberValues(mt, key, expected, actual)
+		if err := compareNumberValues(mt, key, expected, actual); err != nil {
+			return err
+		}
+		return nil
 	case bson.TypeString:
 		val := expected.StringValue()
 		if val == "42" {
-			assert.NotEqual(mt, bson.TypeNull, actual.Type, "expected non-null value for key %v, got null", key)
-			return
+			if actual.Type == bson.TypeNull {
+				return fmt.Errorf("expected non-null value for key %s, got null", key)
+			}
+			return nil
 		}
-		assert.Equal(mt, expected.Value, actual.Value,
-			"value mismatch for key %v; expected %v, got %v", key, expected.Value, actual.Value)
+		break // compare bytes for expected.Value and actual.Value outside of the switch
 	case bson.TypeEmbeddedDocument:
 		e := expected.Document()
 		if typeVal, err := e.LookupErr("$$type"); err == nil {
 			// $$type represents a type assertion
 			// for example {field: {$$type: "binData"}} should assert that "field" is an element with a binary value
-			assertType(mt, actual.Type, typeVal.StringValue())
-			return
+			return assertType(mt, key, actual.Type, typeVal.StringValue())
 		}
 
 		a := actual.Document()
-		compareDocs(mt, e, a)
+		return compareDocsHelper(mt, e, a, key)
 	case bson.TypeArray:
 		e := expected.Array()
 		a := actual.Array()
-		compareDocs(mt, e, a)
-	default:
-		assert.Equal(mt, expected.Value, actual.Value,
-			"value mismatch for key %v; expected %v, got %v", key, expected.Value, actual.Value)
+		return compareDocsHelper(mt, e, a, key)
+	}
+
+	if expected.Type != actual.Type {
+		return fmt.Errorf("type mismatch for key %s; expected %s, got %s", key, expected.Type, actual.Type)
 	}
+	if !bytes.Equal(expected.Value, actual.Value) {
+		return fmt.Errorf("value mismatch for key %s; expected %s, got %s", key, expected.Value, actual.Value)
+	}
+	return nil
 }
 
 // helper for $$type assertions
-func assertType(mt *mtest.T, actual bsontype.Type, typeStr string) {
+func assertType(mt *mtest.T, key string, actual bsontype.Type, typeStr string) error {
```

Comment:
nit: this isn't really an assert anymore

Reply:
changed to `checkValueType`